base/channels.jl

```diff
@@ -543,6 +543,43 @@ function n_avail(c::Channel)
     @atomic :monotonic c.n_avail_items
 end
 
+"""
+    isfull(c::Channel)
+
+Determines if a [`Channel`](@ref) is full, in the
+sense that calling `put!(c, some_value)` will block.
+Returns immediately, does not block.
+
+Note that if another task is currently waiting to `take!`
+a value from `c`, a call to `put!(c, some_value)` might
+immediately be unblocked, even if `isfull(c)==true`.
+This is particularly relevant for unbuffered channels.
```

Comment:
While this looks like a good warning, it seems odd to me to describe the function as "test if full in the sense that `put` will block", and in the next sentence say that "maybie frequently this will ont block when returning true". If I read that, I would be left with a feeling of "okay, so which one is it". In that sense I like my original version, that just explains a case in which it might go wrong, and leaves it to the user to use that knowledge.

But the warning could be harder. What about
> Note that a call to `isfull` generally occurs in a different instant than any call to `put!`. In particular, calling `isfull` and then subsequently `put!` guarantees that the calls occur at different instants. Therefore, the state of the channel may change between the calls, and generally `isfull` can not safely be used to see if a call to `put!` will in fact block.

A little verbose, but very clear and understandable IMO.

Reply:
Your version uses some words that have specific meanings for thread atomics, and those words don't apply in their defined technical sense (in particular "currently" is a term that seems to imply seq-cst guarantees, which does not hold here)

Reply:
The word "instant" is also a seq-cst term, and does not apply here

Reply:
I ment to say a different point in time. Would that be better?

Reply:
In concurrent code with channels, "a point in time" is not a concept that exists

Reply:
Maybe something like "By calling isfull, the caller can establish a happens-before relationship between a prior state of the channel and the future state after a `put!` call on the current thread. By putting a lock on the channel around two calls, or by guaranteeing there is a single producer, the caller can use this to create blocking-free code. However note that this may create livelocks, which may be worse."

Reply:
> In concurrent code with channels, "a point in time" is not a concept that exists

I mean, surely if I start a very accurate watch outside the computer, there is a point in time in which it is decided if the return value is true or false, right? And that clock could form a reference between tasks, which would determine the sequence in which events happen between the tasks? It at least makes inntuitive sense that any event must have a temporal coordinate, just like any object must have a spatial position. But perhaps the issue is how well-defined that point is, like the exact position of a gas cloud?

> "By calling isfull, the caller can establish a happens-before relationship between a prior state of the channel and the future state after a put! call on the current thread. By putting a lock on the channel around two calls, or by guaranteeing there is a single producer, the caller can use this to create blocking-free code. However note that this may create livelocks, which may be worse."

This is not really readable to me, and I find it hard to turn that explanation into an understanding of what goes on, or what I can/can not do...

What about a more generic warning:
> Note that the state of a channel can change between calls to `isfull` and `put!`, meaning that there is no general and safe way to use the two functions to guarantee that a call to `put!` will or will not actually block.

Reply:
I just read my latest suggestion again, and I like it a lot. It is short, precise, and understandable.

Reply:
You are describing seq-cst, which is an ordering that exists, and is the strictest ordering, but is not the ordering that applies here. Events do have temporal coordinates, and they do have a time at that temporal coordinate, but that does not imply anything about the sequence in which events happen at other coordinates in space.

Reply:
I am surely in over my head regarding the details you are describing. If you were able to write something in roughly the format of my latest suggestion, but correct, I think that might be best. Otherwise, your original proposal is alright by me, though harder to read than would be ideal.

Reply:
A minor tweak to you original proposal:
```
Note that it may be frequently the case that `put!` will not block after this returns
`true`. Users must take precautions not to accidentally create live-lock bugs in their code
by calling this method, as these are generally harder to debug than deadlocks. It is also
possible that `put!` will block after this call returns `false`, if there are multiple producer
tasks calling `put!` in parallel.
```

I just aded "returns `false`" in the final sentence. But that is looks grat by me, specially concidering how far into the weeds the concersation goes when I try to be precise. I will commit this version of your original proposal.